nservicebus/handlers/async-handlers.md

```diff
@@ -0,0 +1,163 @@
+
+WARN: It is difficult to give generic advice how asynchronous code should be structured. It is important to understand compute-bound vs. IO-bound as summarized above. Avoid copy-pasteing snippets without further analysis if they provide benefit for the involved business scenarios. Don't assume. Measure it!
+
+
+## Calling short running compute-bound code
+
+Short running compute-bound code (a few milliseconds) that is executed in the handler should be executed directly on the IO-thread that is executing the handler code.
+
+snippet: ShortComputeBoundMessageHandler
+
+Call the code directly and **do not** wrap it with a [`Task.Run`](https://msdn.microsoft.com/en-us/library/system.threading.tasks.task.run.aspx) or [`Task.Factory.StartNew`](https://msdn.microsoft.com/en-au/library/dd321439.aspx).
+
+
+## Calling long-running compute-bound code
+
+Long running compute-bound code (more than a hundred milliseconds) that is executed in a handler could be offloaded to the worker thread pool.
+
+snippet: LongComputeBoundMessageHandler
+
+Wrap the compute-bound code in a [`Task.Run`](https://msdn.microsoft.com/en-us/library/system.threading.tasks.task.run.aspx) or [`Task.Factory.StartNew`](https://msdn.microsoft.com/en-au/library/dd321439.aspx) and `await` the result of the task.
+
+
+## Return or await
+
+
+### Awaits the task
+
+For the majority of cases just mark the handler's Handle method with the `async` keyword and `await` all asynchronous calls inside the method.
+
+snippet: HandlerAwaitsTheTask
+
+
+### Returns the task
+
+For high-throughput scenarios and if there are only one or two asynchronous exit points in the Handle method the `async` keyword can be avoided completely by returning the task instead of awaiting it. This will omit the state machine creation which drives the async code and reduce the number of allocations on the given code path.
+
+snippet: HandlerReturnsATask
+
+snippet: HandlerReturnsTwoTasks
+
+
+## Usage of ConfigureAwait
+
+By default when a task is awaited a mechanism called context capturing is enabled. The current context is captured and restored for the continuation that is scheduled after the precedent task was completed.
+
+snippet: HandlerConfigureAwaitSpecified
+
+In the snippet above `SomeAsyncMethod` and `AnotherAsyncMethod` are simply awaited. So when entering `SomeAsyncMethod` the context is captured and restored before `AnotherAsyncMethod` is executed. The context capturing mechanism is almost never needed in code that is executed inside handlers or sagas. NServiceBus makes sure the context is not captured in the framework at all. So the following approach is preferred:
+
+snippet: HandlerConfigureAwaitNotSpecified
+
+Specify [`ConfigureAwait(false)`](https://msdn.microsoft.com/en-us/library/system.threading.tasks.task.configureawait.aspx) on each awaited statement. Apply this principle to all asynchronous code that is called from handlers and sagas.
+
```

Comment:
Is it worth making a statement about calling `Send()` / `Publish()` outside of a handler. i.e. In a UI